### PR TITLE
[GB][E2E] Add `getTestAccountByFeature` account-selection abstraction for E2E tests

### PR DESCRIPTION
**Pull request by fullofcaffeine**
Follow up to: https://github.com/Automattic/wp-calypso/pull/61349.
Project: p9oQ9f-13V-p2.

### Changes proposed in this Pull Request

This PR aims to abstract the WP account selection logic for our E2E tests in order to make it easier to support more complex scenarios where a specific site is selected based on external data (passed by env var / TC). The original goal was to make it trivial to support testing on AT - with the flip of a switch. I've done that by adding a new `TEST_ON_ATOMIC` TC param and env var that is then passed to a new function that processes this criteria (and others, more about it below) and returns the correct site (in this case, the AT site if the var is set -- and the AT site to be returned might even change based on whether or not the feature has a flag (i.e is Gutenberg `edge` or `stable`?).

#### Introducing the `getTestAccountByFeature` function

This PR adds a helper function to abstract the selection of an accountName for a test. It allows for a declarative way to define and select WP account names for each feature or combination of features (in this specific case, the only "features" available are `gutenberg` and `coblocks`, but we could extend this concept later to selectively select sites for more components or combo of components).

To illustrate, the function simplifies this:

```js
let accountName: string;
if ( envVariables.COBLOCKS_EDGE ) {
	accountName = 'coBlocksSimpleSiteEdgeUser';
} else if ( envVariables.GUTENBERG_EDGE ) {
	accountName = 'gutenbergSimpleSiteEdgeUser';
} else {
	accountName = 'gutenbergSimpleSiteUser';
}
```

Into: 

```js
const accountName = getTestAccountByFeature( {
	coblocks: envVariables.COBLOCKS_EDGE ? 'edge' : undefined,
	gutenberg: envVariables.GUTENBERG_EDGE ? 'edge' : 'stable',
	siteType: envVariables.TEST_ON_ATOMIC ? 'atomic' : 'simple',
} );
```

@worldomonation is working on refactoring some editor classes to support AT. These classes will feed from the account name returned by the new function and the new `TEST_ON_ATOMIC` env var to decide what implementation to use (we should assume AT doesn't use Gutenframe, see: . This is the main difference between the editor on simple vs Atomic at the moment). The changes by @worldomonation **shouldn't prevent this PR from being merged**. The changes here shouldn't break anything, they just allow us to "flip the switch" later by creating new builds to run these tests on AT.

Once these changes are in, we can then create a new build that will pass the `TEST_ON_ATOMIC` var and all the tests that we refactored to use the `getTestAccountByFeature` for the account-selection logic will start selecting the right AT site to test with (provided the actual criteria is present for it!).

From then on, it'll be a matter of seeing what errors and fixing to make them work in AT. Some tests might have to be skipped (we can use the `skipItIf` helper to skip if `TEST_ON_ATOMIC` is set) because they might test something that's too tied to the Gutenframe/Calypso, and that's fine.

### Next Steps

- Add AT-specific Gutenberg upgrade builds in TC that will set the `TEST_ON_ATOMIC` env variable automatically.

### Later

We could explore expanding the concept introduced by this function to conditionally return other artifacts based on different testing criteria. For example, it could return a more complex object with boolean flags in addition to the account name or anything else that needs to be conditional to external vars. Maybe it could be used to also return different components or test implementations based on the env. TBD.

### TODO

- [x] Convert tests that require a test account and have any kind of selection logic around it to use the `getTestAccountByFeature` function instead and define the conditions (criteria) declaratively as needed (in progress)
- [x] Add tests (in progress)
- [x] Add a `AT_TEST` (naming is hard!) parameter to TC and to the `envVariables` in `calypso-e2e`. This will be passed to `getTestSite` to get a proper account based on its presence (besides `GUTENBERG_EDGE`).
- [x] Add jsdoc for `getTestAccountByFeature`
- [x] Fix lint errors
- [x] Rebase and fix conflicts

#### Testing instructions

* Tests should pass

Related to https://github.com/Automattic/wp-calypso/pull/61724, https://github.com/Automattic/wp-calypso/pull/61932

### CODE REVIEW

File: test/e2e/specs/blocks/blocks__smoke-test-across-gb-edge-versions.ts
Comment (fullofcaffeine) [inline test/e2e/specs/blocks/blocks__smoke-test-across-gb-edge-versions.ts]: This is just for illustration purposes, I'll remove the `activationCriteria` soon as this test doesn't need this.

Comment (fullofcaffeine) [reply test/e2e/specs/blocks/blocks__smoke-test-across-gb-edge-versions.ts]: The idea of this param is to allow some tests to activate a target based on external factors (i.e an env var). The gist is that not all tests will need to run on AT all the time.

File: test/e2e/specs/appearance/widgets__legacy-visibility.ts
Comment (fullofcaffeine) [inline test/e2e/specs/appearance/widgets__legacy-visibility.ts]: Ops, `TEst` 🤦🏻

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/utils/get-test-account.ts]: Style suggestion: could we drop the `is`, so it reads simply `edge`: true/false, `atomic`: true/false?

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/utils/get-test-account.ts]: Style suggestion: `pluginName`

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account.ts]: Sure :)

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (fullofcaffeine) [inline packages/calypso-e2e/src/lib/utils/get-test-account.ts]: I guess we could rename the `overridenAccounts`  to `accounts`

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/utils/get-test-account.ts]: Perhaps this name can be a bit more explicit - something like `TestAccountOptions`?

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/utils/get-test-account.ts]: Suggestion: `GutenbergEnvironment`.

If I understand correctly, this is to specify the target for Gutenberg build, correct?

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account.ts]: It could be used for Coblocks, too (see the last exported function).

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/utils/get-test-account.ts]: The presence of this here is bit of a problem, since an account can have multiple sites of which some can be `atomic` and some can be `simple`. 

With that said, I don't have a good solution to this, unless we want to start abstracting sites - but that seems rather excessive, so as long as we stay within the bounds of test accounts this assumption should hold.

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/utils/get-test-account.ts]: I think having dedicated methods that fill in part of the required details (like these) are fine as long as the combinations don't grow too much.

I tend to prefer having a generic method/function that can handle different scenarios, so something like:

```
export function getTestAccountByFeature( plugin: string, options: Options ) {
}

```

At that point it might make sense to rename the `getTestSite` function into `getTestAccountByFeature` so callers don't have to pass though two layers.

This is a stylistic suggestion; the current implementation is perfectly suitable for now, I think.

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account.ts]: Ah, that's a good point! I think we have been following this pattern of one site per account right? Do you know of any editor tests that actually select a different site after login from the top of your head?

File: packages/calypso-e2e/src/lib/utils/get-test-account.ts
Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account.ts]: I like the idea of making it more general and by feature. In the future, we might use this feature to select specific sites for other plugins or even other components. I'll play with it a bit.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (fullofcaffeine) [inline packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: Ops :D

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (fullofcaffeine) [inline packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts:51]: I'd like to clarify the relationship between this `defaultCriteria` and the one that might be passed as the second argument for the `getTestAccountByFeature`. For example, if you look at this test: https://github.com/Automattic/wp-calypso/pull/61783/files#diff-62d89bc54131b1f1bba8fd14ded781bc4a3fd8336a0a479fe1d311eda0b5b994, you'll notice that while it uses the `gutenbergSimpleSiteEdgeUser` as the edge account (which is what we consider the "default" one here), for stable it uses `simpleSitePersonalPlanUser` instead of `gutenbergSimpleSiteEdgeUser`. There are probably good reasons for that, and we should support it. To support that, you'll see that in the aforementioned test I'm passing a custom criterion that replaces the default defined here.

Perhaps having a default that's used implicitly could cause more harm than good? Perhaps it'd be better to always pass a `criteria` to the `getTestAccountByFeature` function inline where it's used in the test? Maybe we could still export this const...

Another possibility is to centralize all the possible criteria here, but for that, we'd probably need an additional attribute, maybe something called `variation` that could have the additional data to make the criteria unique. So for example, we could have:

```typescript
const defaultCriteria: FeatureCriteria[] = [
   {
      "gutenberg":"edge",
      "siteType":"simple",
      "accountName":"gutenbergSimpleSiteEdgeUser"
   },
   {
      "gutenberg":"edge",
      "siteType":"simple",
      "accountName":"gutenbergSimpleSiteUser"
   },
   {
      "gutenberg":"edge",
      "siteType":"simple",
      "variation":"siteEditor",
      "accountName":"siteEditorSimpleSiteUser"
   },
   {
      "gutenberg":"edge",
      "siteType":"simple",
      "variation":"siteEditor",
      "accountName":"siteEditorSimpleSiteEdgeUser"
   }
]
```

We could then use it like this:

```js
const accountName = getTestAccountByFeature({gutenberg: 'edge', siteType: 'simple'});
// returns "siteEditorSimpleSiteUser"
const accountName = getTestAccountByFeature({gutenberg: 'edge', siteType: 'simple', variation: 'siteEditor'});
// returns "siteEditorSimpleSiteEdgeUser'"
```

We could still allow the function to accept a second `criteria` argument to augment the internal map with more options - inline - if needed:

```js
const accountName = getTestAccountByFeature({gutenberg: 'edge', siteType: 'simple', variation: 'newOne'}, [{gutenberg: 'edge', siteType: 'simple', variation: 'newOne', accountName: 'kawabanga'}]);
// returns "kawabanga"
);
```

What do you think?

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts:51]: I implemented support for this. There's now an additional (and optional) `variation` attribute as part of the `FeatureKey` type. We can use that to keep all `feature -> site` definitions (aka criteria) in a central place (probably [here](https://github.com/Automattic/wp-calypso/pull/61783/files#diff-9002489e27625a1d0b7fa59a8f40b5fdafe17d31b263fa5511df77ab88f3edacR34)?).

Comment (worldomonation) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts:51]: Would it make things simpler if we drop the use of `simpleSitePersonalPlanUser`?

The `simpleSiteXYZ` users were introduced some months ago as a way to distribute the testing load to multiple users. However, with discussion of WordPress Plans reorganization and use of preauthenticated cookies we can probably go back to using a smaller number of accounts.

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts:51]: Sure, having fewer accounts/sites is better, but not strictly necessary in this case. I was mainly thinking out loud and documenting how it worked, and I already added support for the additional attribute, so we can have a central list of sites if we want (and avoid [this](https://github.com/Automattic/wp-calypso/pull/61783/files#diff-62d89bc54131b1f1bba8fd14ded781bc4a3fd8336a0a479fe1d311eda0b5b994R37-R42)). The goal of this is to provide a clean and simple way of selecting an account from a variety of possible combinations based on external data passed to the test.

File: test/e2e/specs/editor/editor__page-basic-flow.ts
Comment (fullofcaffeine) [inline test/e2e/specs/editor/editor__page-basic-flow.ts]: I think I've made an oversight here and passing `{ gutenberg: 'stable', siteType: 'simple', accountName: 'gutenbergSimpleSiteEdgeUser' }` is not needed (since it's already defined as part of the [default criteria](https://github.com/Automattic/wp-calypso/pull/61783/files#diff-9002489e27625a1d0b7fa59a8f40b5fdafe17d31b263fa5511df77ab88f3edacR34-R56)). In fact, it'd be better to just have everything defined in there and I think I'll just do it.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (dpasque) [inline packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: This whole implementation is really slick! 💪  This definitely seems like it could be extended really easily!

Is this potentially overkill though for our majority use cases? I'm definitely not saying that's the case, but wanted to open up the discussion. 

The way I see it, the vast majority of scripts are going to do (or could be adjusted to do) one of three things:
1. Just use the default user (this is Calypso-unique flows)
2. Use the same shared set of logic to select a Gutenberg user based on values from `envVariables`
3. Be unique enough that it has its own custom set of users and unique logic to pick them

If that's the case, would it be easier to just make a function that handles the no. 2 case by taking in an `envVariables` object and returning a user?

I wonder if that might simplify things for most test writers because they don't need to repeat all the ternary expressions to create the `FeatureKey` object, or have to worry about making sure there's an exact match of Feature values in the `defaultCriteria` table.

I see two potential benefits of this:
1. A lot of test scripts could just start with a call that looks like...
```typescript
const testAccount = getGutenbergTestAccount( envVariables );
``` 
2. It might make the logical flow a little bit easier to follow. I realize this is likely very much my personal preference, but I find it a little easier to follow a logical flow in statements than matching up all the keys and values in a table. 

Again, I'm not entirely sure I even agree with myself here. 😆  And this current approach is really cool! I just wanted to double check there wasn't an easier option out there! I hope that's not unhelpful!

Comment (dpasque) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: As a middle ground to this too, we could also just add a second function that transforms an `envVariables` object to a `FeatureKey` object, so that we would have to replicate the ternary logic as much in all our Gutenberg scripts.

File: packages/calypso-e2e/test/lib/utils/get-account-by-feature.test.ts
Comment (dpasque) [inline packages/calypso-e2e/test/lib/utils/get-account-by-feature.test.ts:5]: Yay!! Thanks for providing unit tests for this! 💛

**Review by dpasque — COMMENTED** [review]
@worldomonation @fullofcaffeine Finally catching up on everything! I think the implementation here is really slick and totally works, but I just had one question to double check that there isn't a simpler way! I would love to hear your take on it!

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: > Is this potentially overkill though for our majority use cases? I'm definitely not saying that's the case, but wanted to open up the discussion.

Yeah, that crossed my mind! I think we could definitely get away without having such an abstraction for some cases. Right now - before we add the AT test "dimension" - it could indeed be considered overkill. As soon as I started asking the "Is this an AT test" question for each test, things started getting a bit messy, though:

```js
let accountName: string;
if ( envVariables.TEST_ON_ATOMIC ) {
	accountName = 'testonatomicuser';
} else if ( envVariables.GUTENBERG_EDGE ) {
	accountName = 'gutenbergSimpleSiteEdgeUser';
} else {
	accountName = 'gutenbergSimpleSiteUser';
}
```

And soon we'll probably also extend CoBlock's tests to run on AT, and this would become a bit more complex:

```js
let accountName: string;
if ( envVariables.TEST_ON_ATOMIC ) {
 accountName = 'regulartestonatomicuser';
} else {
 if ( envVariables.COBLOCKS_EDGE ) {
 	accountName = 'coBlocksSimpleSiteEdgeUser';
 } else if ( envVariables.GUTENBERG_EDGE ) {
	 accountName = 'gutenbergSimpleSiteEdgeUser';
 } else {
	accountName = 'gutenbergSimpleSiteUser';
 }
}
```

It's possible that we might extend this concept to test a future version of another plugin in AT (less likely, but who knows). In fact, we'll probably have two sites for GB and AT (right now we only have one, for edge) and will allow the `GUTENBERG_EDGE` flag to influence which AT site is selected, so the logic with conditionals would look something like:

```js
let accountName: string;
if ( envVariables.TEST_ON_ATOMIC ) {
  if ( envVariables.GUTENBERG_EDGE ) {
    accountName = 'atgbedgeaccount';
  } else {
   accountName = 'atstablegbaccount';
  }
} else {
 if ( envVariables.COBLOCKS_EDGE ) {
 	accountName = 'coBlocksSimpleSiteEdgeUser';
 } else if ( envVariables.GUTENBERG_EDGE ) {
	 accountName = 'gutenbergSimpleSiteEdgeUser';
 } else {
	accountName = 'gutenbergSimpleSiteUser';
 }
}
```

It could possibly be tidied up with ternary expressions and some helpers locally maybe but at the expense of not being DRY. It's not a huuuge deal, but is a bit painful to be doing this every time and to read such a long (and possibly confusing) chain of conditional statements. That's why I thought a declarative way - something like a table - would be better - it just feels more natural to reason about and write. The "syntax" of each criterion is basically the following question:

* Does the state of the world look like x? Then account = y. `x` being `FeatureKey` and `y` being the `accoutName`. Ex: `{ gutenberg: 'edge', coblocks: 'edge', siteType: 'simple', accountName: 'accountForGBedgeAndCoBlocksEdgeOnSimple' }`.

This has the benefit of being easy to extend for any other scenarios where you need a site to be conditional on external flags (see the `Later` section in the PR description or more ideas on what we could do in the future).

That fits the pattern of a test site per configuration we follow now very naturally. We are basically writing conditions for when this site is selected in a table of objects instead of using conditionals, which in my opinion is easier to read and scan through, with the nice side effect of DRYing up the client code to select the account.

> but I find it a little easier to follow a logical flow in statements than matching up all the keys and values in a table.

For most simple cases where the outcome is based on a single flag (or might not need a flag at all) then I agree. However, the more variables in the equation, the messier the code will become, though. It's possible we might not go beyond adding the `TEST_ON_ATOMIC` variable, but what if have more specific test sites in the future? For example, what if we add a new parameter that passes a theme (for the most important, most used themes) to the test and we then have specific sites for:
* theme x + site editor + gutenberg edge on simple
* theme x + site editor + gutenberg stable on atomic

This would get out of control with `if` conditionals. Having the function in place of a simple conditional in those cases is a small price to pay in order to build an uniform interface for parameterization (see [#](https://github.com/Automattic/wp-calypso/pull/61783#discussion_r827465855)).

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: > const testAccount = getGutenbergTestAccount( envVariables );

I think you make a very good point here! 99% of the time the data will come from the env variables.

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: Another benefit of using such a function is to create a common interface to parameterize tests based on the env variables. I.e using the theme example above. Let's say we want to start running test y in theme x when gutenberg is edge and stable. We can just add a new "theme" string parameter to TC and to the type behind `envVariables`. If we change the function to just accept the env variables object, we'd just need to add new entries to the table (and update the `FeatureKey` type to have a new `theme` attribute of type `string`):

* `{ gutenberg: 'edge', siteType: 'simple', theme: 'themex', accountName: gbEdgeThemeXSimple'}`
* `{ gutenberg: 'stable', siteType: 'simple', theme: 'themex', accountName: 'gbStableThemeXSimple'}`

And then create a build in TC for each theme (I think).

Comment (dpasque) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: You're always so thorough in your explanations, thank you so much for that! 🙌  You have me sold here! 😄  Thanks for thinking far ahead! I agree that the future cases you described would be _so_ much more cleanly handled with the set-up you've created.

Comment (dpasque) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: And I'll leave it up to you whether having middleware that auto-makes the ENV --> `FeatureSet` conversion helps or hurts the readability/maintainability!

**Comment by dpasque** [comment]
@fullofcaffeine  I think we're pretty close to set here! You mentioned wanting to potentially centralize the `FeatureCriteria` tables even more so there's not too many defined out in individual specs, but I saw there were still a few specs that had their custom criteria. Did you want to do that on this PR before merging?

**Comment by fullofcaffeine** [comment]
> @fullofcaffeine I think we're pretty close to set here! You mentioned wanting to potentially centralize the FeatureCriteria tables even more so there's not too many defined out in individual specs, but I saw there were still a few specs that had their custom criteria. Did you want to do that on this PR before merging?

Yes! I've been busy with a related diff today (D75693-code), but I'll try to get to it later today or early tomorrow. That being said, I think it doesn't hurt to leave the second optional `criteria` argument in case there's a need 🤔 I'll ponder a bit more about this.

**Comment by fullofcaffeine** [comment]
@dpasque @worldomonation I think I'll leave the middleware function (discussed [here](https://github.com/Automattic/wp-calypso/pull/61783#discussion_r827379747)) to a subsequent PR (that I can pick up right after merging this). I'll add the other feature criteria to the central table now, and will fix conflicts.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: I think it's worth adding that helper function, but I'll work on it in a follow-up.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (dpasque) [inline packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: Tiny optional style nit I should have mentioned before, but stylistically I always prefer to give these args in callbacks more clear names, even though they're part of commonly known signatures. Can we call this something like `currentMap` or something similar?

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (dpasque) [inline packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: Again, a tiny style nit here, but this name sounds kind of `boolean` to me, could we make it `mergeAndOverrideCriteria`? I know that's long, but might make a little more sense to consumers.

**Review by dpasque — APPROVED** [review]
LGTM! 🚀  I left two tiny style nits -- feel free to do them if you think they're helpful!

I _love_ the new merge-and-override approach, I think that reads really nicely and builds in the perfect amount of flexibility! 👌 

@worldomonation  do you want to take a final look too?

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: Sure! 👍🏻

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts
Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts]: Sounds good! 👍🏻

**Pull request merged**
